Settings page: Replace screen options with tabs

mehmoodak opened the pull request:
## Description
Added tabs in Parse.ly settings and removed screen options.

Closes: #939 

## How has this been done?
- Removes screen options and help text.
- Added tabs.
- Refactors validate functions i.e. divided one big function which now reprsents fields of  each tab.

## What else do I need to know?
- Renamed `Requires Recrawl Settings` to `Recrawl Settings` because of better naming

## How has this been tested?
- Tested manually that each field are storing fine (better if someone else also does this just to be safe)
- Changed existing tests to reflect new changes

## Screenshots

![image](https://user-images.githubusercontent.com/31419912/215747052-2cb8b56f-f78d-4b2e-9ed8-1e6f8558a133.png)

mehmoodak committed:
👌 IMPROVE: replaces screen options with tabs in Parse.ly Settings

mehmoodak commented on src/UI/class-settings-page.php:
```diff
@@ -316,7 +271,7 @@ public function initialize_settings(): void {
 	private function initialize_basic_section(): void {
 		add_settings_section(
 			'basic_settings',
-			__( 'Basic Settings', 'wp-parsely' ),
```
Removed because now tabs already shows this and having title also seems redundant.

mehmoodak commented on src/UI/class-settings-page.php:
```diff
@@ -998,13 +995,16 @@ public function validate_options( $input ) {
+			$input['meta_type'] = $options['meta_type'];
+		} else {
+			$input['meta_type'] = sanitize_text_field( $input['meta_type'] );
+		}
```
Moved this to basic section function.

mehmoodak commented on src/UI/class-settings-page.php:
```diff
@@ -998,13 +995,16 @@ public function validate_options( $input ) {
 		}
 
-		// Validate 'Track post type as'.
-		$this->validate_options_post_type_tracking( $input );
```
Moved this to recrawl settings function

mehmoodak commented on src/UI/class-settings-page.php:
```diff
@@ -1026,18 +1026,6 @@ public function validate_options( $input ) {
-		} else {
-			$input['disable_autotrack'] = 'true' === $input['disable_autotrack'];
-		}
-
```
Moved this to advanced settings function

mehmoodak commented on src/UI/class-settings-page.php:
```diff
@@ -1057,14 +1045,20 @@ public function validate_options( $input ) {
 			$input['disable_amp'] = 'true' === $input['disable_amp'];
 		}
 
-		$input['api_secret'] = sanitize_text_field( $input['api_secret'] );
```
Was a duplicate so removed.

mehmoodak commented on views/parsely-settings.php:
```diff
@@ -11,27 +11,45 @@
+
+	<?php $wp_parsely_settings->show_setting_tabs(); ?>
+
+	<div class="tab-content">
```
Added tabs and wrapped the function inside `.tab-content`

mehmoodak commented on views/parsely-settings.php:
```diff
@@ -11,27 +11,45 @@
+		<form
+			name="parsely"
+			method="post"
+			action=<?php echo esc_url( 'options.php?tab=' . $wp_parsely_settings->get_active_tab() ); ?>
```
Passed tab info so that validations will be run only for the fields in active tab.

mehmoodak commented on wp-parsely.php:
```diff
@@ -134,8 +134,8 @@ function parsely_admin_init_register(): void {
 function parsely_wp_admin_early_register(): void {
-	$settings_page = new Settings_Page( $GLOBALS['parsely'] );
-	$settings_page->run();
+	$GLOBALS['parsely_settings_page'] = new Settings_Page( $GLOBALS['parsely'] );
```
Using global variable because we are using this same instance for showing tabs on `views/parsely-settings.php`

acicovic replied:
What does the transition to the global variable offer us? Was it required to make things work?

mehmoodak replied:
I have to use this same instance in `views/parsely-settings.php` so therefore put it in globals.

acicovic commented:
I absolutely love the screenshot!

mehmoodak reviewed (commented):
Added comments for easier review.

acicovic reviewed (changes requested):
First of all, I'll push this back to 3.8 (as was originally the plan) so we have more time to reflect and discuss this change.

E2E tests constantly fail, so this needs fixing.

While I like the simplicity of the current implementation, the user experience has issues. As every tab is treated as a separate page, we have a page reload when moving between tabs. I've noticed that the current solution:

1. Requires users to click the save button on every tab they change settings in.
2. Does not warn that settings will be lost when changes are performed but are not saved and the user navigates to another tab. Users do not expect their changes to be lost as they navigate between tabs.

Most of the tab implementations I've seen do not treat separate tabs as different pages, and I think that this relates to the UX issues above.

So my take would be:
- Let's aim to do this without needing page reloads between tabs.
- If we conclude for any reason that we do not want to go that route, we should notify the user if their form is dirty and they try to navigate to another tab.

mehmoodak commented:
> E2E tests constantly fail, so this needs fixing.

It was intentional i.e. trying to get first review and then target fixing of e2e tests.

mehmoodak committed:
Merge branch 'develop' into 939-add-tabs-for-parsely-screen-options

mehmoodak committed:
👌 IMPROVE: print each section field separately

mehmoodak committed:
👌 IMPROVE: tabs UX

mehmoodak committed:
➕ ADDS: section titles

mehmoodak committed:
🐛 FIX: integration tests

mehmoodak commented:
@acicovic  feedback implemented about changing tabs without reload. Please have a look again. Thanks

FYI: Will fix e2e tests soon just wanted to get first review before fixing.

mehmoodak committed:
🐛 FIX: e2e tests

mehmoodak commented:
e2e tests are also updated.

mehmoodak committed:
👌 IMPROVE: sonar cloud duplication issue

dependabot[bot] committed:
build(deps-dev): bump @typescript-eslint/eslint-plugin

Bumps [@typescript-eslint/eslint-plugin](https://github.com/typescript-eslint/typescript-eslint/tree/HEAD/packages/eslint-plugin) from 5.52.0 to 5.53.0.
- [Release notes](https://github.com/typescript-eslint/typescript-eslint/releases)
- [Changelog](https://github.com/typescript-eslint/typescript-eslint/blob/main/packages/eslint-plugin/CHANGELOG.md)
- [Commits](https://github.com/typescript-eslint/typescript-eslint/commits/v5.53.0/packages/eslint-plugin)

---
updated-dependencies:
- dependency-name: "@typescript-eslint/eslint-plugin"
  dependency-type: direct:development
  update-type: version-update:semver-minor
...

Signed-off-by: dependabot[bot] <support@github.com>

acicovic committed:
Merge pull request #1408 from Parsely/dependabot/npm_and_yarn/typescript-eslint/eslint-plugin-5.53.0

acicovic committed:
README.md: Update contributor list

acicovic commented:
Hello, I'll start reviewing this. Can you resolve the conflict?

acicovic commented on src/UI/class-settings-page.php:
```diff
@@ -284,6 +203,45 @@ public function display_settings(): void {
+		<nav class="nav-tab-wrapper">
+			<?php foreach ( $this->setting_tabs as $t ) { ?>
+			<a
+				class="nav-tab <?php echo esc_attr( $t['key'] . '-tab' ); ?>"
```
At least for PHP code, my suggestion is to rename `$t` to `$tab` (for here and everywhere else this is done). Makes things more obvious/clearer.

acicovic commented on src/UI/class-settings-page.php:
```diff
@@ -284,6 +203,45 @@ public function display_settings(): void {
+	 *
+	 * @since 3.8.0
+	 */
+	public function show_content_of_setting_tabs(): void {
```
Could be renamed to `show_setting_tabs_content()` (I think it's better but you can decide).

acicovic commented on src/UI/class-settings-page.php:
```diff
@@ -475,16 +431,18 @@ private function initialize_basic_section(): void {
 			__( 'Requires Recrawl Settings', 'wp-parsely' ),
 			function (): void {
-				echo '<strong>' . wp_kses_post( __( '<span style="color:#d63638">Important:</span> Changing any of these values below on a site currently tracked with Parse.ly will require reprocessing of your Parse.ly data.', 'wp-parsely' ) ) . '</strong><br />';
+				echo '<br /><strong>' . wp_kses_post( __( '<span style="color:#d63638">Important:</span> Changing any of these values below on a site currently tracked with Parse.ly will require reprocessing of your Parse.ly data.', 'wp-parsely' ) ) . '</strong><br />';
```
I cannot see this text in the UI, and I think it should be there. Also, why is this `<br>` needed?

mehmoodak replied:
1. Added the text in UI
2. Added `<br>` to improve spacing else text appears right below the tabs which looks a bit odd

acicovic commented on src/UI/class-settings-page.php:
```diff
@@ -475,16 +431,18 @@ private function initialize_basic_section(): void {
 	 * @since 3.2.0
 	 */
-	private function initialize_requires_recrawl_section(): void {
+	private function initialize_recrawl_section( string $section_key ): void {
```
If we keep this renaming, maybe it's worth adding a since comment that the method has been renamed int 3.8.0.

acicovic replied:
Marking this as resolved since we will be keeping the new naming.

acicovic commented on src/UI/class-settings-page.php:
```diff
@@ -1058,14 +1023,20 @@ public function validate_options( $input ) {
-			$input['meta_type'] = sanitize_text_field( $input['meta_type'] );
-		}
+	/**
+	 * Validate fields of Recrawl Section.
```
```suggestion
	 * Validates fields of Recrawl Section.
```

acicovic commented on src/UI/class-settings-page.php:
```diff
@@ -1163,6 +1134,31 @@ public function validate_options( $input ) {
 	}
 
+	/**
+	 * Validate fields of Advanced Section.
```
```suggestion
	 * Validates fields of Advanced Section.
```

acicovic reviewed (commented):
@mehmoodak, this is great! The user experience is much smoother in my opinion!

Did a first review, leaving some some comments and questions.

@danielabloch, this PR renames the `Requires Recrawl Settings` section to `Recrawl Settings`. Is this OK with you, provided we have the text that explains that a recrawl is required after setting changes?

Also, I wonder if we want to remove the `Settings` string on every tab (see screenshot below - may be rendundant?). If we're unsure, we can leave as is though.

![image](https://user-images.githubusercontent.com/23142906/220337170-43708225-8444-43ee-8fd2-dc8c45dcc1bd.png)

acicovic committed:
Merge pull request #1412 from Parsely/update/readme.md-contributors

mehmoodak committed:
Refactor permission checks to hide Parse.ly widget and stats column based on user capabilities (#1407)

* Adds checks on endpoints

* Fix integration tests

danielabloch commented:
> this PR renames the Requires Recrawl Settings section to Recrawl Settings. Is this OK with you, provided we have the text that explains that a recrawl is required after setting changes?

> Also, I wonder if we want to remove the Settings string on every tab (see screenshot below - may be rendundant?). If we're unsure, we can leave as is though.


1. Yes, Recrawl Settings is a good change, still makes sense. 
2. Agreed, we should provide text explaining that recrawl is required after a setting changes.
3. I also think Settings string on every tab is redundant.

dependabot[bot] committed:
build(deps): bump release-drafter/release-drafter from 5.22.0 to 5.23.0

Bumps [release-drafter/release-drafter](https://github.com/release-drafter/release-drafter) from 5.22.0 to 5.23.0.
- [Release notes](https://github.com/release-drafter/release-drafter/releases)
- [Commits](https://github.com/release-drafter/release-drafter/compare/v5.22.0...v5.23.0)

---
updated-dependencies:
- dependency-name: release-drafter/release-drafter
  dependency-type: direct:production
  update-type: version-update:semver-minor
...

Signed-off-by: dependabot[bot] <support@github.com>

acicovic committed:
Merge pull request #1414 from Parsely/dependabot/github_actions/release-drafter/release-drafter-5.23.0

acicovic committed:
Add update version number (uvn) script

acicovic committed:
uvn.sh: Auto-add and auto-commit

acicovic committed:
Merge pull request #1416 from Parsely/add/update-version-number-script

mehmoodak committed:
Add filters to configure user capability on remote APIs (#1417)

* ➕ ADDS: filter on remote apis to change user capability

* ➕ ADDS: integration tests

mehmoodak committed:
🐛 FIX: recrawl notice and improve comments

mehmoodak committed:
Merge branch develop into 939-add-tabs-for-parsely-screen-options

mehmoodak committed:
💫 UPDATE: readme and screenshots

mehmoodak commented:
Feedback incorporated and screenshots for plugins directory are also updated.

Please have a look again. Thanks

acicovic reviewed (commented):
I'd remove the version strings from the screenshots, because they currently convey inaccurate information (the feature is not available in 3.6.2) and they will make our screenshots look outdated very soon.

Apart from that, everything looks good to me 👍

Let's not merge this yet though, as the decision to add this to 3.7 or 3.8 is pending.

mehmoodak committed:
💫 UPDATE: screenshots

acicovic reviewed (approved):
Looks good. Thanks for bearing with me 🙂

Approving this, but let's not merge it yet please.

acicovic commented:
Scheduling this for 3.8. I've changed the base branch, so it can be merged.
